src/Microsoft.DotNet.Build.Tasks.Templating/GenerateFileFromTemplate.cs

```diff
@@ -0,0 +1,147 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace Microsoft.DotNet.Build.Tasks.Templating
+{
+    /// <summary>
+    /// <para>
+    /// Generates a new file at <see cref="OutputPath"/>.
+    /// </para>
+    /// <para>
+    /// The <see cref="TemplateFile"/> can define variables for substitution using <see cref="Properties"/>.
+    /// </para>
+    /// <example>
+    /// The input file might look like this:
+    /// <code>
+    /// 2 + 2 = ${Sum}
+    /// </code>
+    /// When the task is invoked like this, it will produce "2 + 2 = 4"
+    /// <code>
+    /// &lt;GenerateFileFromTemplate Properties="Sum=4;OtherValue=123;" ... &gt;
+    /// </code>
+    /// </example>
+    /// </summary>
+    public class GenerateFileFromTemplate : Task
+    {
+        /// <summary>
+        /// The template file.
+        /// Variable syntax: ${VarName}
+        /// If your template file needs to output this format, you can escape the dollar sign with a backtick, e.g. `${NotReplaced}
+        /// </summary>
+        [Required]
+        public string TemplateFile { get; set; }
+
+        /// <summary>
+        /// The destination for the generated file.
+        /// </summary>
+        [Required]
+        [Output]
+        public string OutputPath { get; set; }
+
+        /// <summary>
+        /// Key=value pairs of values
+        /// </summary>
+        [Required]
+        public string[] Properties { get; set; }
+
+        public override bool Execute()
+        {
+            var outputPath = Path.GetFullPath(OutputPath.Replace('\\', '/'));
+
+            if (!File.Exists(TemplateFile))
+            {
+                Log.LogError($"File {TemplateFile} does not exist");
+                return false;
+            }
+
+            var values = MSBuildListSplitter.GetNamedProperties(Properties);
+            var template = File.ReadAllText(TemplateFile);
+
+            var result = Replace(template, values);
+            Directory.CreateDirectory(Path.GetDirectoryName(outputPath));
+            File.WriteAllText(outputPath, result);
+
+            return true;
```

Comment:
nit: this pattern helps guard against errors-but-task-succeeded cases:
```suggestion
            return !Log.HasLoggedErrors;
```